src/async_wrap.cc

```diff
@@ -670,6 +673,10 @@ void AsyncWrap::AsyncReset(Local<Object> resource, double execution_async_id,
                                                      : execution_async_id;
   trigger_async_id_ = env()->get_default_trigger_async_id();
 
+  auto isolate = Isolate::GetCurrent();
+  resource_.Reset(isolate, resource);
+  resource_.SetWeak();
```

Comment:
I'm curious if there is a chance that the underlying resource object gets GCed before it's used in `before` hook (read in `InternalCallbackScope`)?

Reply:
I agree that this might happen.
See e.g. https://github.com/nodejs/node/pull/30196 for a similar problem which was fixed by adding a strong reference from the underlying resource to the wrapper.
I think we should agree on where to put this strong reference - once for all in async_wrap or at the place a wrapper is created.

Reply:
Yeah, this is tricky – on the one hand, it should be a strong reference, because otherwise this might end up being GC’ed too early, on the other hand the `resource` object might point back to the bindings object (aka `object()`), preventing it from being GC’ed when it’s a weak object (like a zlib stream.)

It should be okay to add a property on the underlying resource object that points to the public wrapper object to solve this; And, as I understand it, the semantics that we’d want here are basically that `[owner_symbol]` on the underlying resource points to the public wrapper object, so maybe we can just use that?

Reply:
At least in the case above (#30196) you added a new symbol.
I took a fast look into the code there and it seems `[owner-symbol]` of handle (a `JSStream` or `TCP`) there points already to a `JSStreamSocket` or `Socket`.

Reply:
Actually, if I just switched it back to being a strong reference again, wouldn't that be fine? The reference from the AsyncWrap instance would only keep it alive as long as the AsyncWrap instance itself was alive and had not yet been reset to another resource yet. I don't see any problem with that. :thinking:

Reply:
I think it depends on which object is kept alive and what else this object might reference. What's wrong with clearing the reference in `AsyncWrap::EmitDestroy()`?

Reply:
The suggestion from @Flarna seems to have fixed it.